File: backend/app/services/student_orgs.py
from app.db.supabase import get_supabase
from app.models.student_orgs import STUDENT_ORGS
from app.schemas.student_orgs import StudentOrgOut


def _map_club_record(record: dict) -> StudentOrgOut:
    return StudentOrgOut(
        id=record["id"],
        name=record["name"],
        abbreviation=record["abbreviation"],
        description=record["description"],
        website=record.get("website"),
        instagram=record.get("instagram"),
        category=record.get("category"),
    )


def _list_student_orgs_from_supabase() -> list[StudentOrgOut]:
    response = (
        get_supabase()
        .table("clubs")
        .select("id, name, abbreviation, description, website, instagram, category")
        .order("name")
        .execute()
    )

    return [_map_club_record(record) for record in (response.data or [])]


def list_student_orgs() -> list[StudentOrgOut]:
    try:
        clubs = _list_student_orgs_from_supabase()
        if clubs:
            return clubs
    except Exception:
        pass

    return [StudentOrgOut(**org) for org in STUDENT_ORGS]

File: frontend/src/lib/studentOrgs.ts
import { studentOrgs as mockStudentOrgs, type StudentOrg } from '@/data/mockData'
import { supabase } from '@/lib/supabase'

interface ClubRecord {
  id: string
  name: string
  abbreviation: string
  description: string
  website: string | null
  instagram: string | null
}

function mapClubRecord(record: ClubRecord): StudentOrg {
  return {
    id: record.id,
    name: record.name,
    abbreviation: record.abbreviation,
    description: record.description,
    website: record.website,
    instagram: record.instagram,
  }
}

export async function getStudentOrgs(): Promise<StudentOrg[]> {
  const { data, error } = await supabase
    .from('clubs')
    .select('id, name, abbreviation, description, website, instagram')
    .order('name', { ascending: true })

  if (error) {
    console.warn('Falling back to bundled mock student orgs because Supabase clubs could not be loaded.', error)
    return mockStudentOrgs
  }

  if (!data?.length) {
    console.warn('Supabase returned no clubs. Falling back to bundled mock student orgs.')
    return mockStudentOrgs
  }

  return data.map(mapClubRecord)
}
